### PR TITLE
feat(app-deploy): support deploy with dockerfile

### PR DESCRIPTION
**Pull request by nettoclaudio**
This PR implements the changes required to support deploy with Dockerfile (and files). 

Significant changes:
* Adds a new parameter `--dockerfile` on `tsuru app-deploy`;
* Changes the default compression level to best (9) - more CPU usage during the deploy/build;
* Rewrites the command description of `tsuru app-deploy` and `tsuru app-build`;
* Shows more info messages while compressing files to `archive.tar.gz`;

### CODE REVIEW

**Review by tcarreira — CHANGES_REQUESTED** [review]

File: tsuru/client/build.go [hunk 1]
```diff
@@ -110,22 +131,31 @@ func (c *AppBuild) Run(context *cmd.Context, client *cmd.Client) error {
 	return cmd.ErrAbortCommand
 }
 
-func uploadFiles(context *cmd.Context, filesOnly bool, request *http.Request, buf *safe.Buffer, body *safe.Buffer, values url.Values) error {
+func uploadFiles(context *cmd.Context, filesOnly bool, request *http.Request, buf *safe.Buffer, body *safe.Buffer, values url.Values, archive io.Reader) error {
```
[the remark this answers is not in the document]
Comment (nettoclaudio) [reply tsuru/client/build.go:134]: Sure... dropping it.

File: tsuru/client/archiver.go [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+	var added int
+
+	for _, path := range paths {
+		abs, err := filepath.Abs(path)
+		if err != nil {
+			return fmt.Errorf("failed to get the absolute filename of %q: %w", path, err)
+		}
+
+		if !strings.HasPrefix(abs, workingDir) {
```
[the remark this answers is not in the document]
Comment (nettoclaudio) [reply tsuru/client/archiver.go:113]: Nice catch! Changing it.

File: tsuru/client/archiver.go [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+		fmt.Fprintf(a.stderr, "WARNING: Skipping file %q due to unsupported file type.\n", filename)
+		return 0, nil
+	}
+
+	if isDir && filesOnly { // there's no need to create dirs in files only
+		return 0, nil
+	}
+
+	if a.ignore.MatchesPath(filename) {
+		fmt.Fprintf(a.stderr, "File %q matches with some pattern provided in the ignore file... skipping it.\n", filename)
```
[the remark this answers is not in the document]
Comment (nettoclaudio) [reply tsuru/client/archiver.go:162]: Yeah, I agree... So I'll take advantage of the `tsuru -v <number>` argument to define if it's running in debug mode. Do you see any problem about this approach?

[the remark this answers is not in the document]
Comment (tcarreira) [reply tsuru/client/archiver.go:162]: I think that's a very good approach 👍

**Review by tcarreira — APPROVED** [review]
Fantastic ❤️

**Pull request merged**